Implement node stop/start/restart for Power

gitsridhar opened the pull request:
This is an implementation of node stop/start/restart for
IBM Power environment other than libvirt. In libvirt
environment, the same is implemented using virsh stop/start etc.
For IBM Power environment other than Libvirt, the same is
attempted with stop/start of kubelet service inside the nodes.

Signed-off-by: gitsridhar <svenkat@us.ibm.com>

vasukulkarni commented on ocs_ci/utility/powernodes.py:
```diff
@@ -56,6 +57,37 @@ def verify_machine_is_down(self, node):
         else:
             return True
 
+    def stop_powernodes_machines_powervs(
```
Instead of power class, it can be its own class `kube_service.py` , I think in general its applicable for all platforms and should be different category than node reboots.

gitsridhar replied:
Vasu, thanks for your comments. But I am aware of only code plugged into platform_nodes. If I implement this into a separate class, how do I link it to platform_nodes? Test cases get an instance of platform node and invoke node restarts from platform_nodes interface.

vasukulkarni replied:
I will check the test case  to see how we can keep the distinction clear.

shylesh replied:
@gitsridhar ,
If I understand your requirements properly, you need control over node and service. Why can't we have a composition here. 
```python

class Service(object):
   self.service_type=None

class Kubelet(Service):
  def __init__(self):
    super().__init__("kubelet")
 
  def start():
  def stop():
  def kill(graceful=True):     

class IBMPowerNodes(NodeBase):
   
  def __init_(self): 
    service = None

  def init_service_obj(self, service_name):
     service = service_class()
```

From the test we can do:
```python
node = IBMPowerNodes()
node.init_service_obj('kubelet')
node.service.start()
node.service.stop()
node.service.kill()
```

We can rethink the class in case if you need multiple service types to be controlled.

Not sure if this helps in your case .
Thoughts ?

shylesh replied:
@gitsridhar , sorry I didn't realize that you are already doing something similar.

petr-balogh replied:
@gitsridhar I agree with Vasu to move this to some other file as stop/restart/start of service can be used not just on power nodes.

So you can move just parts from here about start/stop/restart to some service.py file and describe that this action happens via ssh.  The rest of the logic can stay in this file as you have.

gitsridhar replied:
Petr and Shylesh, thanks for your quick comments. I will amend this PR shortly based on your comments.

I am doing this based on one observation: when I manually ssh into a storage node and stop kubelet service, the node moves to the NotReady state and in about 5 minutes, all pods running on that node get terminated and either relocated to another node or moves to the pending state. When kubelet service is started, pods get back to the original Ready state. I understand this is equivalent to shutting down the entire storage node and starting it back after few minutes by ocs-ci disruptive test cases. Please comment if this understanding is not correct.

gitsridhar replied:
@shylesh I am doing only part of what you said (tie into IBMPowerNodes..) but I agree with you on class separation, so it can be used by anyone for any service.

gitsridhar replied:
@shylesh @petr-balogh @vasukulkarni I just pushed an amend to PR, I have addressed all of your comments.

petr-balogh commented on ocs_ci/utility/powernodes.py:
```diff
@@ -56,6 +57,37 @@ def verify_machine_is_down(self, node):
+            powernode_machines (list): PowerNodes objects
+            timeout (int): time in seconds to wait for node to reach 'not ready' state
+            wait (bool): True if need to wait till the restarted node reaches timeout
+                         - for future use
```
too big indention here, only 4 spaces please

petr-balogh commented on ocs_ci/utility/powernodes.py:
```diff
@@ -90,6 +122,34 @@ def stop_powernodes_machines(
+        """
+        for node in powernode_machines:
+            result = exec_cmd(
+                f"ssh core@{node.name} sudo systemctl start kubelet.service"
```
I think that node.name is not reliable value for IP address as this is not always IP address but can be also the regular name.   Node IP can be taken from status:
```
status:
  addresses:
  - address: 10.243.128.6
    type: InternalIP
  - address: 10.243.128.6
    type: ExternalIP
  - address: 10.243.128.6
    type: Hostname
```
This is other part which should be moved to service.py and you will just import it and call it here.

So I would go with ExternalIP type.

We have this function for that purpose:
https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/ocs/node.py#L251

So I think you should use this and get one of external IP to be used here. So maybe you can write new function in node.py which will be called get_node_ip which will return only one reachable IP.

gitsridhar replied:
@petr-balogh I used InternalIP and submitted an amendment to PR. Let me know if there are any concerns.

petr-balogh commented on ocs_ci/utility/powernodes.py:
```diff
@@ -56,6 +57,37 @@ def verify_machine_is_down(self, node):
+            UnexpectedBehaviour: If PowerNode machine is still up
+        """
+        for node in powernode_machines:
+            cmd = f"ssh core@{node.name} sudo systemctl stop kubelet.service"
+            result = exec_cmd(cmd)
```
So basically this will be moved out of here to service.py module.  And I would then also check the real status of the service there if it's really stopped.

gitsridhar reviewed (commented):
I tried to gather ExternalIP for node to use it instead of node name. Here is what I found in my OCP 4.6 environment. I do not see ExternalIP. I see only InternalIP. I am planning to use the InternalIP.

status:
  addresses:
  - address: 10.17.120.156
    type: InternalIP
  - address: master0.sridhar-ocp-ocs.cp.fyre.ibm.com
    type: Hostname
  allocatable:

@petr-balogh Can you comment on this?

gitsridhar commented:
My latest PR amendment has an implementation of Service and KubeletService classes which can be re-used by any other module in OCS-CI.

vasukulkarni commented on ocs_ci/utility/powernodes.py:
```diff
@@ -90,6 +119,23 @@ def stop_powernodes_machines(
+            timeout (int): time in seconds to wait for node to reach 'not ready' state,
+                and 'ready' state.
+            wait (bool): Wait for PowerNodes to start - for future use
+            force (bool): True for PowerNode ungraceful power off, False for
```
instead of 'force', can you call this `using_method=kube-service`, I think that will give users who are reading code better insights into what method is used for disruption.  You will have to fix at few more places below.

gitsridhar replied:
@vasukulkarni I see your point, but the parameter force is used as a boolean value and specifically, in stop methods, its boolean value is used to add '-f' parameter to systemctl command. I have enhanced the logging in server.py to include service name (like kubelet.service) so it will be very clear on which component and service is invoking the code.

vasukulkarni commented on ocs_ci/utility/service.py:
```diff
@@ -0,0 +1,245 @@
+        of nodes with node name as keys.
+        """
+        self.service_name = service_name
+        self.force = force
```
I think you can get rid this option in this class,  The class doesn't need to be aware of this parameter.

vasukulkarni replied:
unless this `force` option is different from what is mentioned above.

gitsridhar replied:
Vasu, see my comment above, the force is used as a boolean parameter.

vasukulkarni commented on ocs_ci/ocs/platform_nodes.py:
```diff
@@ -1736,8 +1736,8 @@ def stop_nodes(self, nodes, force=True):
-            raise NotImplementedError(
-                "This is not libvirt environment. Stop nodes not implemented"
+            self.powernodes.stop_powernodes_machines_powervs(
+                nodes, timeout=900, wait=True, force=force
```
`use_method=kube-service` is better option than `force`

gitsridhar replied:
Vasu, see my other comments above.

vasukulkarni replied:
@gitsridhar I see there are 2 force options, I meant changing of the the force of stop_power and start_power functions to use `use_method` to distinguish between service and real power restarts

gitsridhar replied:
@vasukulkarni I have simplified the service, There are no more two different 'force' parameters. There is only one now when the service object is initialized, the force parameter can be initialized.  Hope this clears the confusion over the force parameter.

gitsridhar commented:
@vasukulkarni @petr-balogh @clacroix12 @shylesh  I have updated the code based on your comments and tested and removed the 'WIP' prefix as well. Please review and merge if everything is ok. This work will help any component of ocs-ci to use systemctl and control any service on any master or worker nodes.

gitsridhar commented:
@vasukulkarni @petr-balogh @clacroix12 @shylesh  Please take a look and merge if everything is ok?

clacroix12 commented on ocs_ci/utility/powernodes.py:
```diff
@@ -56,6 +61,30 @@ def verify_machine_is_down(self, node):
+        waiting_time = 360
+        logger.info(f"Waiting for {waiting_time} seconds")
+        time.sleep(waiting_time)
+        logger.info(f"Waiting over for {waiting_time} seconds")
```
This log message is similar to the one before and makes it seem like we're waiting again. I would make this a bit more clear("Done waiting" or "Continuing...") or simply remove this log statement entirely.

gitsridhar replied:
Done. I removed the second log parameter.

clacroix12 commented on ocs_ci/utility/service.py:
```diff
@@ -0,0 +1,247 @@
+
+    The purpose of this class is to extend Service class to provide stop/start/restart etc operations on
+    kubelet service. Since kubelet service stop and start operations involve changing status of OCP node
+    obects, this class verifies the same.
```
```suggestion
    objects, this class verifies the same.
```

gitsridhar replied:
Done.

clacroix12 reviewed (commented):
This looks great! Really like the separation of the generic service function from the extra verification done in the `KubeletService` class. I just had a few minor comments but nothing that should hold this PR up.

gitsridhar commented on ocs_ci/ocs/node.py:
```diff
@@ -278,6 +278,30 @@ def get_node_ips(node_type="worker"):
+        ipkind: ExternalIP or InternalIP or Hostname
+
+    Returns:
+        dictionary: Internal or Exteranl IP addresses keyed off of node name
```
Done.

vasukulkarni commented on ocs_ci/utility/service.py:
```diff
@@ -0,0 +1,242 @@
+            timeout (int): time in seconds to wait for service to stop.
+
+        """
+        super().stop(node, timeout)
```
The implementation looks reversed,  we might actually not need either parent or child class, since you are doing heavy lifting in parent and waiting in child.  I will approve the PR not to hold it further.

gitsridhar replied:
I will revisit this in the next iteration. I feel that parent class should do as much in a generic way and child class should do whatever is specific to themselves. I will initiate a discussion when we get time to revisit this design.
